### DISCUSSION TITLE
Use of Compute shaders / GPU

### DISCUSSION THREAD

**Discussion by Deytooh** [naverz/zepeto-studio-global#17] [Q&A]
Hi, 

I was wondering if ComputeBuffer were working for zepeto in typescript ?
I tried with some types but I cant make work ComputeBuffer.SetData method, this method in C# would require blittable types, but in typescript I dont know which types it requires, so I keep getting unvalid arguments exception.

**Comment by Deytooh** [comment]
So actually blittable types are okay, but for some reasons ComputeBuffer.Setdata keeps throwing an error, so I tried to use ComputeBuffer.BeginWrite but it says that "BeginWrite is not a function" so...

**Comment by jhslee** [comment]
Hello again @Deytooh! 👋 

I am currently in the process of looking into your issue. May I take a look at a sample code for your ComputeBuffer.SetData method, your BeginWrite implementation, as well as a snippet of the buffer in the shader you're trying to write to? It would help me to try and track this issue down 😁

**Reply by Deytooh** [reply]
Hi again @jhslee 🔥 
Yeah sure !! 
Here is the script ->
This is actually just the script from Unity https://docs.unity3d.com/ScriptReference/Graphics.DrawMeshInstancedIndirect.html
I wanted to see if we could use DrawMeshInstancedIndirect on Zepeto :) 
The problem comes from the type that we set to the method ComputeBuffer.SetData, I tried with many different types but still doesnt work, even the List$1 couldnt make the work :( Do you have any solution ? If we can use DrawMeshInstancedIndirect it would be really interesting for pushing the limits of Zepeto :)

`import { List$1 } from 'System.Collections.Generic';
import { Bounds, ComputeBuffer, ComputeBufferMode, ComputeBufferType, Graphics, GraphicsBuffer, Input, Material, Mathf, Mesh, Random, Vector3, Vector4 } from 'UnityEngine';
import { ZepetoScriptBehaviour } from 'ZEPETO.Script'
export default class InstanceExm extends ZepetoScriptBehaviour {

    public instanceCount: int = 100000;
    public instanceMesh: Mesh;
    public instanceMaterial: Material;
    public subMeshIndex: int = 0;

    private cachedInstanceCount: int = -1;
    private cachedSubMeshIndex: int = -1;
    private positionBuffer: ComputeBuffer;
    private argsBuffer: ComputeBuffer;
    private args: int[] = [0, 0, 0, 0, 0];
    private argsCs: List$1<int>;

    private mat: Material;

    Start() {
        
        this.cachedInstanceCount = -1;
        this.cachedSubMeshIndex = -1;
        this.subMeshIndex = 0;
        this.instanceCount = 100000;
        this.argsCs = new List$1<int>();//<int>(5) doesnt work...
         for (let i = 0; i < 5; i++) {
             this.argsCs.Add(0);
         }
        console.log(this.argsCs.Count);
        this.argsBuffer = new ComputeBuffer(1, this.argsCs.Count * 4, ComputeBufferType.IndirectArguments);
        this.UpdateBuffers();

    }

    Update() {
        // Update starting position buffer
        if (this.cachedInstanceCount != this.instanceCount || this.cachedSubMeshIndex != this.subMeshIndex)
            this.UpdateBuffers();

        // Render
        Graphics.DrawMeshInstancedIndirect(this.instanceMesh, this.subMeshIndex, this.instanceMaterial, new Bounds(Vector3.zero, new Vector3(100.0, 100.0, 100.0)), this.argsBuffer);
    }


    UpdateBuffers() {
        // Ensure submesh index is in range
        if (this.instanceMesh != null)
            this.subMeshIndex = Mathf.Clamp(this.subMeshIndex, 0, this.instanceMesh.subMeshCount - 1);

        // Positions
        if (this.positionBuffer != null)
            this.positionBuffer.Release();
        this.positionBuffer = new ComputeBuffer(this.instanceCount, 16, ComputeBufferType.IndirectArguments, ComputeBufferMode.SubUpdates);

        let positions = new List$1<Vector4>(); // List$1<Vector4>(this.instanceCount) doesnt work so...
        for (let i = 0; i < this.instanceCount; i++) {
            let angle = Random.Range(0.0, Mathf.PI * 2.0);
            let distance = Random.Range(20.0, 100.0);
            let height = Random.Range(-2.0, 2.0);
            let size = Random.Range(0.05, 0.25);
            positions.Add(new Vector4(Mathf.Sin(angle) * distance, height, Mathf.Cos(angle) * distance, size));
        }

        this.positionBuffer.SetData(positions);
        this.instanceMaterial.SetBuffer("positionBuffer", this.positionBuffer);

        console.log(this.instanceCount);
        // Indirect args
        if (this.instanceMesh != null) {
            this.argsCs[0] = this.instanceMesh.GetIndexCount(this.subMeshIndex);
            this.argsCs[1] = this.instanceCount;
            this.argsCs[2] = this.instanceMesh.GetIndexStart(this.subMeshIndex);
            this.argsCs[3] = this.instanceMesh.GetBaseVertex(this.subMeshIndex);
        }
        else {
            this.argsCs[0] = this.argsCs[1] = this.argsCs[2] = this.argsCs[3] = 0;
        }

        this.argsBuffer.SetData(this.argsCs);

        this.cachedInstanceCount = this.instanceCount;
        this.cachedSubMeshIndex = this.subMeshIndex;
    }

    OnDisable() {
        if (this.positionBuffer != null)
            this.positionBuffer.Release();
        this.positionBuffer = null;

        if (this.argsBuffer != null)
            this.argsBuffer.Release();
        this.argsBuffer = null;
    }
}

`

**Reply by Deytooh** [reply]
You can get the shader that you need to use fort the property "instanceMaterial" on the Unity page of DrawMeshInstancedIndirect :) and instanceMesh you can set any mesh, but for performances reasons, a quad will be enough !

**Reply by Deytooh** [reply]
So I got helped from Puerts developer, and by changing "ComputeBufferType.IndirectArguments" to 256 hard coded, and this.argsBuffer.SetData(this.argscs.ToArray()); and this.positionBuffer.SetData(positions.ToArray());  there is no more error throwed by the SetData :) But the DrawMeshInstancedIndirect still does not work xD

**Reply by Deytooh** [reply]
Okay I got to work DrawMeshInstancedIndirect by filling out every parameters ( even if we dont need everything actually ) but it displays nothing so mhh... quite stuck there ! Here is my final code ->
Does it displays something on your side ?

`  public instanceCount: int = 100000;
    public instanceMesh: Mesh;
    public instanceMaterial: Material;
    public subMeshIndex: int = 0;

    private cachedInstanceCount: int = -1;
    private cachedSubMeshIndex: int = -1;
    private positionBuffer: ComputeBuffer;
    private argsBuffer: ComputeBuffer;
    private args: int[] = [0, 0, 0, 0, 0];
    private argsCs: List$1<int>;
    private mat: Material;

    Start() {
        // const CS = $require('csharp');

        this.cachedInstanceCount = -1;
        this.cachedSubMeshIndex = -1;
        this.subMeshIndex = 0;
        this.instanceCount = 100000;
        this.argsCs = new List$1<int>();//<int>(5) doesnt work...
         for (let i = 0; i < 5; i++) {
             this.argsCs.Add(0);
         }
        console.log(this.argsCs.Count);
        this.argsBuffer = new ComputeBuffer(1, this.argsCs.Count * 4,0);
        this.UpdateBuffers();

    }

    Update() {
        // Update starting position buffer
        if (this.cachedInstanceCount != this.instanceCount || this.cachedSubMeshIndex != this.subMeshIndex)
            this.UpdateBuffers();

        // Render
        Graphics.DrawMeshInstancedIndirect(this.instanceMesh, this.subMeshIndex, this.instanceMaterial, new Bounds(Vector3.zero, new Vector3(100.0, 100.0, 100.0)), this.argsBuffer,
        0,null,1,true,0,null,0,null);
    }


    UpdateBuffers() {
        // Ensure submesh index is in range
        if (this.instanceMesh != null)
            this.subMeshIndex = Mathf.Clamp(this.subMeshIndex, 0, this.instanceMesh.subMeshCount - 1);

        // Positions
        if (this.positionBuffer != null)
            this.positionBuffer.Release();
        this.positionBuffer = new ComputeBuffer(this.instanceCount, 16, 256);

        let positions = new List$1<Vector4>(); // List$1<Vector4>(this.instanceCount) doesnt work so...
        for (let i = 0; i < this.instanceCount; i++) {
            let angle = Random.Range(0.0, Mathf.PI * 2.0);
            let distance = Random.Range(20.0, 100.0);
            let height = Random.Range(-2.0, 2.0);
            let size = Random.Range(0.05, 0.25);
            positions.Add(new Vector4(Mathf.Sin(angle) * distance, height, Mathf.Cos(angle) * distance, size));
        }

        this.positionBuffer.SetData(positions.ToArray());
        this.instanceMaterial.SetBuffer("positionBuffer", this.positionBuffer);

        console.log(this.instanceCount);
        // Indirect args
        if (this.instanceMesh != null) {
            this.argsCs[0] = this.instanceMesh.GetIndexCount(this.subMeshIndex);
            this.argsCs[1] = this.instanceCount;
            this.argsCs[2] = this.instanceMesh.GetIndexStart(this.subMeshIndex);
            this.argsCs[3] = this.instanceMesh.GetBaseVertex(this.subMeshIndex);
        }
        else {
            this.argsCs[0] = this.argsCs[1] = this.argsCs[2] = this.argsCs[3] = 0;
        }

        this.argsBuffer.SetData(this.argsCs.ToArray());
        

        this.cachedInstanceCount = this.instanceCount;
        this.cachedSubMeshIndex = this.subMeshIndex;
    }

    OnDisable() {
        if (this.positionBuffer != null)
            this.positionBuffer.Release();
        this.positionBuffer = null;

        if (this.argsBuffer != null)
            this.argsBuffer.Release();
        this.argsBuffer = null;
    }`

**Reply by Deytooh** [reply]
Okay, made it work :) 
Thank you for your help still xD !! I hope it can help other people also.
If someone needs it , here is how to make works DrawMeshInstancedIndirect in typescript ->
( you can get the shader from the public unity wikipage )
```
public instanceCount: int = 100000;
    public instanceMesh: Mesh;
    public instanceMaterial: Material;
    public subMeshIndex: int = 0;

    private cachedInstanceCount: int = -1;
    private cachedSubMeshIndex: int = -1;
    private positionBuffer: ComputeBuffer;
    private argsBuffer: ComputeBuffer;
    private args: int[] = [0, 0, 0, 0, 0];
    private argsCs: List$1<int>;
    private mat: Material;

    Start() {
        // const CS = $require('csharp');

        this.cachedInstanceCount = -1;
        this.cachedSubMeshIndex = -1;
        this.subMeshIndex = 0;
        this.instanceCount = 100000;
        this.argsCs = new List$1<int>();//<int>(5) doesnt work...
         for (let i = 0; i < 5; i++) {
             this.argsCs.Add(0);
         }
        console.log(this.argsCs.Count);
        this.argsBuffer = new ComputeBuffer(1, this.argsCs.Count * 4,256);
        this.UpdateBuffers();

    }

    Update() {
        // Update starting position buffer
        if (this.cachedInstanceCount != this.instanceCount || this.cachedSubMeshIndex != this.subMeshIndex)
            this.UpdateBuffers();

        // Render
        Graphics.DrawMeshInstancedIndirect(this.instanceMesh, this.subMeshIndex, this.instanceMaterial, new Bounds(Vector3.zero, new Vector3(100.0, 100.0, 100.0)), this.argsBuffer,
        0,null,1,true,0,null,0,null);
    }


    UpdateBuffers() {
        // Ensure submesh index is in range
        if (this.instanceMesh != null)
            this.subMeshIndex = Mathf.Clamp(this.subMeshIndex, 0, this.instanceMesh.subMeshCount - 1);

        // Positions
        if (this.positionBuffer != null)
            this.positionBuffer.Release();
        this.positionBuffer = new ComputeBuffer(this.instanceCount, 16, 256);

        let positions = new List$1<Vector4>(); // List$1<Vector4>(this.instanceCount) doesnt work so...
        for (let i = 0; i < this.instanceCount; i++) {
            let angle = Random.Range(0.0, Mathf.PI * 2.0);
            let distance = Random.Range(20.0, 100.0);
            let height = Random.Range(-2.0, 2.0);
            let size = Random.Range(0.05, 0.25);
            positions.Add(new Vector4(Mathf.Sin(angle) * distance, height, Mathf.Cos(angle) * distance, size));
        }

        this.positionBuffer.SetData(positions.ToArray());
        this.instanceMaterial.SetBuffer("positionBuffer", this.positionBuffer);

        console.log(this.instanceCount);
        // Indirect args
        if (this.instanceMesh != null) {
            this.argsCs[0] = this.instanceMesh.GetIndexCount(this.subMeshIndex);
            this.argsCs[1] = this.instanceCount;
            this.argsCs[2] = this.instanceMesh.GetIndexStart(this.subMeshIndex);
            this.argsCs[3] = this.instanceMesh.GetBaseVertex(this.subMeshIndex);
        }
        else {
            this.argsCs[0] = this.argsCs[1] = this.argsCs[2] = this.argsCs[3] = 0;
        }

        this.argsBuffer.SetData(this.argsCs.ToArray());
        

        this.cachedInstanceCount = this.instanceCount;
        this.cachedSubMeshIndex = this.subMeshIndex;
    }

    OnDisable() {
        if (this.positionBuffer != null)
            this.positionBuffer.Release();
        this.positionBuffer = null;

        if (this.argsBuffer != null)
            this.argsBuffer.Release();
        this.argsBuffer = null;
    }
```

**Reply by jhslee** [reply]
This is wonderful!! 😍 Thanks @Deytooh, for sharing your solution and i'm so glad you were able to figure it out! This will be a huge resource and inspiration to any technical artists and creators looking to push the boundaries of what Zepeto is capable of.

**Reply by Deytooh** [reply]
Yeah sure !! Thank you !! 💯  I really hope it will be useful for everyone struggling on this ! 🔥 It would be nice to make a post in the korean channel also ;)
